(maybe) Fix failing tests for BCF XML importer

cbliard opened the pull request:
There is something going on with validity_period.
The new code seems to work.

The tests are failing since the modifications done in #21325.

We paired on it with @ulferts. Can you help further please?

cbliard committed:
[69448] Send assignee notification email only once after PIR completion

https://community.openproject.org/wp/69448

In `WorkPackages::CreateService`, the work package was saved twice:
first time at creation, second time when rescheduling related work
packages, which includes self.

Each save created a `WorkflowJob`. Normally this is ok, but if there are
mentions in the description or journal_notes, and if the two jobs are
processed in parallel (in different GoodJob threads), this leads to
having the mention notification email sent twice. Indeed, each one will
try to create the notifications and send them. The `mail_alert_sent`
flag on the notification is not set to `true` until one mail has been
sent.

To fix it, the work packages that have not been changed after
rescheduling related work packages are filtered out. This guarantees
there is only one single save in the nominal cases where there are no
related work packages that would force a second save of the work
package. Single save -> single mention notification email.

cbliard committed:
Fix failing tests for BCF XML importer

There is something going on with validity_period.
The new code seems to work.

cbliard committed:
Add failing test for adding multiple comments right after creation

cbliard commented:
@ulferts I added a failing test reproducing the issue in `spec/models/work_package/work_package_acts_as_journalized_spec.rb:60`.

cbliard commented:
Another category of tests is also failing, related to project copy: `./spec/workers/copy_project_job_spec.rb:83`

Before the modifications done in #21325, there was two saves:
- a first one in the service `#perform` with `work_package.save`
- a second one in the same `#perform` when calling `#reschedule_related`: the work package would be saved a second time, and errors stored if the save fails

After the modifications done in #21325, there is only one save:
- in the service `#perform` with `work_package.save`
- the second one in `#reschedule_related` is not triggered because the work package was not modified by the rescheduling (that's the modifications from #21325), so no second save

Too bad, that's on the second save that the errors were merged to the result. Without the second save, and ignoring the errors from the first save, the job completes without errors, but the test fails because it expects some errors.

That's why the test was passing, and now is not passing anymore. But there is more!

Actually, the code should not have tried to reschedule at all because the first save did not work.
There was a subtle issue introduced in [this commit related to automatic subject](https://github.com/opf/openproject/commit/07002998bf77d3be69bec06721939cb344ea5257#diff-d4f369d238b7cb53b2c607b16068605cd4e844e5caa99cff4233dcb251dfdd4fL55-L59)
- before the `result.success` was updated with the result of `work_package.save`
- now the `result.success` is updated with the result of `set_templated_subject(work_package)` which always returns `true` if there is no replacement pattern.

That means that `result.success` is almost always `true` regardless of the work package being valid or not.

Now why is it reaching the first `#save` point?

It's because contract is `WorkPackages::CopyProjectContract` and this one skips model validation (done in [this commit](https://github.com/opf/openproject/commit/a25c70822a7f0084c404092f1e0ddac0bc09a78e)), so the `SetAttributesService` does not raise any error.

And why is the work package eventually saved?

It's because the `WorkPackages::UpdateAncestorsService` is being called too, and this one saves with `validate: false` in `#save_updated_work_packages`.

That's a mess.

And I'm not sure what the expected behavior is. If a work package has some errors on save, should it be saved anyway when copying a project?

cbliard commented:
Ok, so we had the opportunity to explore it more with @EinLama.
Here is what we found about the current behavior on latest `dev` when copying a project:
- a work package with a model error (like `done_ratio` being 101) will be copied, and the error will be reported in the copy.
- a work package with some contract errors will be copied if the `CopyProjectContract` did override the validation for it (for instance, work package is automatically scheduled but has no predecessor nor children). The error will **not** be reported and the work package will be created in the destination project.
- a work package with some contract errors will not be copied if the `CopyProjectContract` has some validation error (like work / remaining work and % complete being incorrect). The error will be reported and the work package will **not** be created in the copied project.
  - furthermore, if that work package had children or relation in the parent project, as the parent is not copied in the destination project, the copied children will still reference the parent of the source project. This feels like a bug.
- we did not try with autogenerated subject, but looking at the code, we suspect that the subject will not be updated if the work package has some model errors. It will simply be the same as the source work package (which will be unnoticed most of the time)

The behavior is not consistent.

To make it consistent, one approach could be:
- to always copy the work package in the destination project, despite the validation errors
- and to always report validation errors in the copy result

oliverguenther commented:
@cbliard I feel like this is more a fundamental question of how errors are being kept and/or recorded in case of copying. I am leaning towards "if you copy the project, you copy as is and accept if the state was faulty" rather than preventing the copy in the first place. At the same time, I think the decisions that lead to these failures are rather arbitrary, so I could see us fixing it in both ways

cbliard commented:
Thanks @oliverguenther 

> I am leaning towards "if you copy the project, you copy as is and accept if the state was faulty" rather than preventing the copy in the first place. 

Same. So we should copy everything if we can, and keep the faulty state.

> At the same time, I think the decisions that lead to these failures are rather arbitrary, so I could see us fixing it in both ways.

Not sure I get what you means.

One thing I'm thinking about: if the copy is done, and is exact, it's probably not necessary to indicate there is an error, because actually if the copy has been done successfully then there is no error. There could be an error only if the copy was impossible.

cbliard committed:
Fix the double save issue

cbliard committed:
Remove debug statement

cbliard committed:
Fix test

cbliard committed:
Use `#save_journals` on Project rather than `#touch_and_save_journals`

`#touch_and_save_journals` seems to have been introduced because of the
bug producing the error "ERROR:  new row for relation "journals"
violates check constraint "journals_validity_period_not_empty"
(PG::CheckViolation)".

As this bug has been fixed, using `#touch_and_save_journals` is not
needed anymore, except if we really want to touch the project
`updated_at`, in which case a unit test should fail.

This partially reverts what was done in 7b93b60.

cbliard committed:
Project copy copies work packages regarless of them being valid or not

When doing a project copy, copy as-is and accept that the state can be
faulty.

Before, it was disabling parts of the validations while keeping some
other parts, which made the experience confusing: the
SetAttributesService would be ok, but then saving the work package would
fail, and later the work package would be saved again without validation
in the `UpdateAncestorsService`.

Actually it was working previously only due to side-effects of
`UpdateAncestorsService` which does save and also due to changes
introduced in 0700299 which refactored the part that detect `save`
returned `false`, leading to believe that `save` always returned `true`.

Now the validation are done solely by the `SetAttributesService`, and
when saving the work package validations are ignored.

cbliard committed:
Fix test

cbliard committed:
Fix rubocop warning

ulferts commented:
@cbliard , with your initial fix on the journaling carried out in a separate PR and my rework of journaling in #21590, I wonder how much we still want to salvage from this PR. The change to `save_journals` rather than the `touch_and_save_journals` comes to mind. The discussion about the project copying seems to be independent from this all??? Do you have an opinion here?

cbliard commented:
@ulferts 

> with your initial fix on the journaling carried out in a separate PR and my rework of journaling in #21590, I wonder how much we still want to salvage from this PR.

Actually I forgot I had to review the PR you wrote. It was not on my radar. Sorry.

> The change to `save_journals` rather than the `touch_and_save_journals` comes to mind. 

Yes, but it can be left as-is too.

> The discussion about the project copying seems to be independent from this all???

Yes, and it's already fixed.

And I just understood that by ".. my rework of journaling in #21590, I wonder how much we still want to salvage from this PR.", this PR refers to _this PR_ #21332 and not #21590.

Yes, it can be closed. I'll do it now.
